hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java

```diff
@@ -1099,34 +1108,6 @@ boolean hasAnyParentEntityNames(BatchIdentifier batchIdentifier) {
 			boolean hasAnyChildEntityNames(BatchIdentifier batchIdentifier) {
 				return childEntityNames.contains( batchIdentifier.getEntityName() );
 			}
-
-			/**
-			 * Check if the this {@link BatchIdentifier} has a parent or grand parent
-			 * matching the given {@link BatchIdentifier reference.
-			 *
-			 * @param batchIdentifier {@link BatchIdentifier} reference
-			 *
-			 * @return This {@link BatchIdentifier} has a parent matching the given {@link BatchIdentifier reference
-			 */
-			boolean hasParent(BatchIdentifier batchIdentifier) {
-				return (
-					parent == batchIdentifier
-					|| parentEntityNames.contains( batchIdentifier.getEntityName() )
-					|| ( parentEntityNames.contains( batchIdentifier.getRootEntityName() ) && !this.getEntityName().equals( batchIdentifier.getRootEntityName() ) )
-					|| parent != null && parent.hasParent( batchIdentifier, new ArrayList<>() )
-				);
-			}
```

Comment:
The above logic is difficult to comprehend because of various corner cases. It would be excellent to get rid of them if a simpler alternative exists.

Reply:
I think the root cause why we still need to rely `parentEntityNames` in face of `parent` is we need multiple `parents`, not single one. It is a graph structure, not a tree structure. Topological sorting algorithm is working fine and accommodate multiple parents.